File: CASSIA_R/inst/python/model_settings.py
"""
Simplified model settings for CASSIA.

Usage:
    from model_settings import resolve_model_name

    # Use tier shortcuts
    model, provider = resolve_model_name("best", "openai")      # -> ("gpt-5.1", "openai")
    model, provider = resolve_model_name("balanced", "anthropic") # -> ("claude-sonnet-4-5", "anthropic")
    model, provider = resolve_model_name("fast", "openrouter")    # -> ("google/gemini-2.5-flash", "openrouter")

    # Or use exact model names
    model, provider = resolve_model_name("gpt-4o", "openai")      # -> ("gpt-4o", "openai")
"""

import json
from typing import Dict, Tuple, Optional
from pathlib import Path


# Valid tier shortcuts
VALID_TIERS = {"best", "balanced", "fast", "recommended"}

# Valid providers
VALID_PROVIDERS = {"openai", "anthropic", "openrouter"}


class ModelSettings:
    """Simple model settings manager."""

    def __init__(self, config_path: Optional[str] = None):
        """
        Initialize ModelSettings.

        Args:
            config_path: Path to model_settings.json. If None, uses default location.
        """
        if config_path is None:
            current_dir = Path(__file__).parent
            possible_paths = [
                current_dir / "data" / "model_settings.json",
                current_dir / "model_settings.json",
            ]
            for path in possible_paths:
                if path.exists():
                    config_path = path
                    break
            if config_path is None:
                config_path = current_dir / "data" / "model_settings.json"

        self.config_path = Path(config_path)
        self.settings = self._load_settings()

    def _load_settings(self) -> Dict:
        """Load settings from JSON file."""
        try:
            with open(self.config_path, 'r', encoding='utf-8') as f:
                return json.load(f)
        except (FileNotFoundError, json.JSONDecodeError):
            return self._get_fallback_settings()

    def _get_fallback_settings(self) -> Dict:
        """Fallback settings if JSON file is not available."""
        return {
            "providers": {
                "openai": {
                    "best": "gpt-5.1",
                    "balanced": "gpt-4o",
                    "fast": "gpt-5-mini",
                    "recommended": "gpt-5.1"
                },
                "anthropic": {
                    "best": "claude-opus-4-5",
                    "balanced": "claude-sonnet-4-5",
                    "fast": "claude-haiku-4-5",
                    "recommended": "claude-sonnet-4-5"
                },
                "openrouter": {
                    "best": "anthropic/claude-sonnet-4.5",
                    "balanced": "openai/gpt-5.1",
                    "fast": "google/gemini-2.5-flash",
                    "recommended": "anthropic/claude-sonnet-4.5"
                }
            }
        }

    def resolve_model_name(self, model_name: str, provider: str) -> Tuple[str, str]:
        """
        Resolve model name to actual model string.

        Args:
            model_name: Either a tier ("best", "balanced", "fast", "recommended")
                       or an exact model name
            provider: Provider name ("openai", "anthropic", "openrouter")

        Returns:
            Tuple of (resolved_model_name, provider)

        Examples:
            >>> resolve_model_name("best", "openai")
            ("gpt-5.1", "openai")
            >>> resolve_model_name("gpt-4o", "openai")
            ("gpt-4o", "openai")
        """
        if not model_name:
            raise ValueError("Model name cannot be empty")

        if not provider:
            raise ValueError("Provider must be specified (openai, anthropic, or openrouter)")

        provider = provider.lower()
        model_name_lower = model_name.lower()

        if provider not in VALID_PROVIDERS:
            raise ValueError(f"Unknown provider: {provider}. Must be one of: {VALID_PROVIDERS}")

        # Check if it's a tier shortcut
        if model_name_lower in VALID_TIERS:
            provider_settings = self.settings.get("providers", {}).get(provider, {})
            resolved = provider_settings.get(model_name_lower)
            if resolved:
                return resolved, provider
            else:
                raise ValueError(f"Tier '{model_name}' not found for provider '{provider}'")

        # Otherwise, return the model name as-is
        return model_name, provider

    def get_available_tiers(self) -> list:
        """Get list of available tier shortcuts."""
        return list(VALID_TIERS)

    def get_available_providers(self) -> list:
        """Get list of available providers."""
        return list(VALID_PROVIDERS)

    def get_model_for_tier(self, tier: str, provider: str) -> str:
        """
        Get the model name for a specific tier and provider.

        Args:
            tier: One of "best", "balanced", "fast", "recommended"
            provider: One of "openai", "anthropic", "openrouter"

        Returns:
            Model name string
        """
        model, _ = self.resolve_model_name(tier, provider)
        return model

    def print_available_models(self):
        """Print all available models in a readable format."""
        print("\n=== Available Models ===\n")
        print("Tiers: best, balanced, fast, recommended\n")

        providers = self.settings.get("providers", {})
        for provider, tiers in providers.items():
            print(f"{provider.upper()}:")
            for tier, model in tiers.items():
                print(f"  {tier:12} -> {model}")
            print()


# Global instance
_model_settings = None


def get_model_settings() -> ModelSettings:
    """Get the global ModelSettings instance."""
    global _model_settings
    if _model_settings is None:
        _model_settings = ModelSettings()
    return _model_settings


def resolve_model_name(model_name: str, provider: str) -> Tuple[str, str]:
    """
    Resolve model name to actual model string.

    Args:
        model_name: Either a tier ("best", "balanced", "fast", "recommended")
                   or an exact model name
        provider: Provider name ("openai", "anthropic", "openrouter")

    Returns:
        Tuple of (resolved_model_name, provider)

    Examples:
        >>> resolve_model_name("best", "openai")
        ("gpt-5.1", "openai")
        >>> resolve_model_name("balanced", "anthropic")
        ("claude-sonnet-4-5", "anthropic")
        >>> resolve_model_name("fast", "openrouter")
        ("google/gemini-2.5-flash", "openrouter")
        >>> resolve_model_name("gpt-4o", "openai")  # exact name
        ("gpt-4o", "openai")
    """
    return get_model_settings().resolve_model_name(model_name, provider)


def get_recommended_model(provider: str) -> Tuple[str, str]:
    """
    Get the recommended model for a provider.

    Args:
        provider: Provider name ("openai", "anthropic", "openrouter")

    Returns:
        Tuple of (model_name, provider)
    """
    return get_model_settings().resolve_model_name("recommended", provider)


def print_available_models():
    """Print all available models."""
    get_model_settings().print_available_models()

File: CASSIA_R/R/model_settings.R
# Model Settings Functions for CASSIA R Package

#' Get Python CASSIA Module
#'
#' @return Python CASSIA module (all functions available at package level)
#' @keywords internal
.get_py_model_settings <- function() {
  # Use the py_cassia from package namespace if available
  # All model settings functions are exported at the CASSIA package level

  # First check if py_cassia exists in the package namespace
  pkg_env <- tryCatch(asNamespace("CASSIA"), error = function(e) NULL)
  if (!is.null(pkg_env) && exists("py_cassia", envir = pkg_env)) {
    py_obj <- get("py_cassia", envir = pkg_env)
    if (!is.null(py_obj)) {
      return(py_obj)
    }
  }

  # Also check parent frames (for interactive use / devtools::load_all)
  py_obj <- get0("py_cassia", envir = parent.frame(), ifnotfound = NULL)
  if (!is.null(py_obj)) {
    return(py_obj)
  }

  # Check in global environment as fallback
  if (exists("py_cassia", envir = .GlobalEnv)) {
    py_obj <- get("py_cassia", envir = .GlobalEnv)
    if (!is.null(py_obj)) {
      return(py_obj)
    }
  }

  # Otherwise, try to import directly
  tryCatch({
    py_cassia_local <- reticulate::import_from_path("CASSIA", path = system.file("python", package = "CASSIA"))
    return(py_cassia_local)
  }, error = function(e) {
    warning("Could not load CASSIA Python package: ", e$message)
    return(NULL)
  })
}

#' Resolve Model Name
#'
#' Resolve a user-provided model name to the actual provider model name.
#' This function automatically maps simplified names (like "gpt4", "claude", "gemini") 
#' to their actual provider-specific names.
#'
#' @param model_name Character string of the model name (can be alias or actual name)
#' @param provider Character string of the provider name (if NULL, will try to infer from model_name)
#'
#' @return A list with two elements: 'model' (actual model name) and 'provider' (provider name)
#' @export
#'
#' @examples
#' \dontrun{
#' # Using simple names
#' resolve_model_name("gpt4")  # Returns gpt-4o with openai provider
#' resolve_model_name("claude")  # Returns claude-3-5-sonnet-latest with anthropic provider
#' resolve_model_name("gemini")  # Returns google/gemini-2.5-flash with openrouter provider
#' 
#' # Using aliases
#' resolve_model_name("sonnet")  # Returns claude-3-5-sonnet-latest
#' resolve_model_name("deepseek")  # Returns deepseek/deepseek-chat-v3-0324
#' 
#' # With specific provider
#' resolve_model_name("gpt-4o", "openai")
#' }
resolve_model_name <- function(model_name, provider = NULL) {
  tryCatch({
    py_settings <- .get_py_model_settings()
    result <- py_settings$resolve_model_name(model_name, provider)
    
    # Convert Python tuple to R list
    return(list(
      model = result[[1]],
      provider = result[[2]]
    ))
  }, error = function(e) {
    warning("Error resolving model name: ", e$message)
    return(list(model = model_name, provider = provider))
  })
}

#' Get Recommended Model
#'
#' Get the recommended model for a provider or use case.
#'
#' @param provider Character string of the provider name (if NULL, returns overall best model)
#' @param use_case Character string of the use case (annotation, scoring, annotation_boost, merging)
#'
#' @return A list with two elements: 'model' (recommended model name) and 'provider' (provider name)
#' @export
#'
#' @examples
#' \dontrun{
#' # Get overall best model
#' get_recommended_model()
#' 
#' # Get best model for specific use case
#' get_recommended_model(use_case = "annotation")
#' get_recommended_model(use_case = "scoring")
#' get_recommended_model(use_case = "annotation_boost")
#' 
#' # Get recommended model for specific provider
#' get_recommended_model(provider = "openai")
#' get_recommended_model(provider = "anthropic")
#' }
get_recommended_model <- function(provider = NULL, use_case = NULL) {
  tryCatch({
    py_settings <- .get_py_model_settings()
    result <- py_settings$get_recommended_model(provider, use_case)
    
    # Convert Python tuple to R list
    return(list(
      model = result[[1]],
      provider = result[[2]]
    ))
  }, error = function(e) {
    warning("Error getting recommended model: ", e$message)
    return(list(model = "google/gemini-2.5-flash", provider = "openrouter"))
  })
}

#' Get Model Information
#'
#' Get detailed information about a model.
#'
#' @param model_name Character string of the model name
#' @param provider Character string of the provider name (if NULL, will try to infer)
#'
#' @return A list with model information or NULL if not found
#' @export
#'
#' @examples
#' \dontrun{
#' # Get info for specific models
#' get_model_info("gpt-4o")
#' get_model_info("claude-3-5-sonnet-latest")
#' get_model_info("google/gemini-2.5-flash")
#' 
#' # Using aliases
#' get_model_info("gemini")
#' get_model_info("deepseek")
#' }
get_model_info <- function(model_name, provider = NULL) {
  tryCatch({
    py_settings <- .get_py_model_settings()
    result <- py_settings$get_model_info(model_name, provider)
    
    if (is.null(result)) {
      return(NULL)
    }
    
    # Convert Python dict to R list
    return(reticulate::py_to_r(result))
  }, error = function(e) {
    warning("Error getting model info: ", e$message)
    return(NULL)
  })
}

#' List Available Models
#'
#' List available models with optional filtering.
#'
#' @param provider Character string to filter by provider
#' @param cost_tier Character string to filter by cost tier (very_low, low, medium, high)
#' @param use_case Character string to filter by use case
#'
#' @return A data frame with model information
#' @export
#'
#' @examples
#' \dontrun{
#' # List all models
#' list_models()
#' 
#' # Filter by provider
#' list_models(provider = "openai")
#' list_models(provider = "anthropic")
#' list_models(provider = "openrouter")
#' 
#' # Filter by cost tier
#' list_models(cost_tier = "low")
#' list_models(cost_tier = "high")
#' 
#' # Filter by use case
#' list_models(use_case = "annotation")
#' list_models(use_case = "scoring")
#' 
#' # Combine filters
#' list_models(provider = "openrouter", cost_tier = "low")
#' }
list_models <- function(provider = NULL, cost_tier = NULL, use_case = NULL) {
  tryCatch({
    py_settings <- .get_py_model_settings()
    result <- py_settings$list_models(provider, cost_tier, use_case)
    
    # Convert Python list of dicts to R data frame
    if (length(result) == 0) {
      return(data.frame(
        name = character(0),
        provider = character(0),
        description = character(0),
        cost_tier = character(0),
        recommended = logical(0),
        stringsAsFactors = FALSE
      ))
    }
    
    # Extract information from Python result
    df <- data.frame(
      name = sapply(result, function(x) x$name),
      provider = sapply(result, function(x) x$provider),
      description = sapply(result, function(x) x$info$description),
      cost_tier = sapply(result, function(x) x$info$cost_tier),
      recommended = sapply(result, function(x) x$info$recommended),
      stringsAsFactors = FALSE
    )
    
    return(df)
  }, error = function(e) {
    warning("Error listing models: ", e$message)
    return(data.frame())
  })
}

#' Get Use Case Recommendations
#'
#' Get recommendations for a specific use case.
#'
#' @param use_case Character string of the use case name
#'
#' @return A list with recommendations
#' @export
#'
#' @examples
#' \dontrun{
#' # Get recommendations for different use cases
#' get_use_case_recommendations("annotation")
#' get_use_case_recommendations("scoring")
#' get_use_case_recommendations("annotation_boost")
#' get_use_case_recommendations("merging")
#' }
get_use_case_recommendations <- function(use_case) {
  tryCatch({
    py_settings <- .get_py_model_settings()
    result <- py_settings$get_use_case_recommendations(use_case)
    
    # Convert Python dict to R list
    return(reticulate::py_to_r(result))
  }, error = function(e) {
    warning("Error getting use case recommendations: ", e$message)
    return(list())
  })
}

#' Print Model Recommendations
#'
#' Print model recommendations in a user-friendly format.
#'
#' @param use_case Character string of the specific use case to show recommendations for (optional)
#'
#' @return None (prints to console)
#' @export
#'
#' @examples
#' \dontrun{
#' # Print all recommendations
#' print_model_recommendations()
#' 
#' # Print recommendations for specific use case
#' print_model_recommendations("annotation")
#' print_model_recommendations("scoring")
#' }
print_model_recommendations <- function(use_case = NULL) {
  tryCatch({
    py_settings <- .get_py_model_settings()
    py_settings$print_model_recommendations(use_case)
  }, error = function(e) {
    warning("Error printing model recommendations: ", e$message)
  })
}

#' Get Model Aliases
#'
#' Get all aliases for a model.
#'
#' @param model_name Character string of the model name
#' @param provider Character string of the provider name (if NULL, will try to infer)
#'
#' @return A character vector of aliases
#' @export
#'
#' @examples
#' \dontrun{
#' # Get aliases for models
#' get_model_aliases("gpt-4o")
#' get_model_aliases("claude-3-5-sonnet-latest")
#' get_model_aliases("google/gemini-2.5-flash")
#' }
get_model_aliases <- function(model_name, provider = NULL) {
  tryCatch({
    py_settings <- .get_py_model_settings()
    result <- py_settings$get_model_aliases(model_name, provider)
    
    # Convert Python list to R vector
    return(as.character(result))
  }, error = function(e) {
    warning("Error getting model aliases: ", e$message)
    return(character(0))
  })
}

#' Get Cost Tier Models
#'
#' Get models in a specific cost tier.
#'
#' @param cost_tier Character string of the cost tier name
#'
#' @return A character vector of model names
#' @export
#'
#' @examples
#' \dontrun{
#' # Get models by cost tier
#' get_cost_tier_models("very_low")
#' get_cost_tier_models("low")
#' get_cost_tier_models("medium")
#' get_cost_tier_models("high")
#' }
get_cost_tier_models <- function(cost_tier) {
  tryCatch({
    py_settings <- .get_py_model_settings()
    result <- py_settings$get_cost_tier_models(cost_tier)
    
    # Convert Python list to R vector
    return(as.character(result))
  }, error = function(e) {
    warning("Error getting cost tier models: ", e$message)
    return(character(0))
  })
}